eval.go

```diff
@@ -624,218 +324,402 @@ func newUserTypeInfo(userType reflect.Type) (*userTypeInfo, error) {
 		return attrInfo{
-			kind:  kindFloat,
-			ftype: field.Type,
-			index: field.Index,
-			asString: func(v reflect.Value) string {
-				return strconv.FormatFloat(v.FieldByIndex(field.Index).Float(), 'g', -1, 64)
+			asString: func(v reflect.Value) (string, bool) {
+				return strconv.FormatFloat(v.FieldByIndex(field.Index).Float(), 'g', -1, 64), true
+			},
+			asBytes: func(v reflect.Value) ([]byte, bool) {
+				return strconv.AppendFloat(nil, v.FieldByIndex(field.Index).Float(), 'g', -1, 64), true
 			},
-			asBytes: func(v reflect.Value) []byte {
-				return strconv.AppendFloat(nil, v.FieldByIndex(field.Index).Float(), 'g', -1, 64)
+			asFloat: func(v reflect.Value) (float64, error) {
+				return v.FieldByIndex(field.Index).Float(), nil
 			},
 		}, nil
+	case reflect.Struct:
+		if field.Type == timeType {
+			return attrInfo{
+				asFloat: func(v reflect.Value) (float64, error) {
+					return float64(v.FieldByIndex(field.Index).Interface().(time.Time).UnixMilli()) / 1000, nil
+				},
+			}, nil
+		}
+		return attrInfo{}, fmt.Errorf("user value field %s has unsupported type %s", field.Name, field.Type)
 	default:
 		return attrInfo{}, fmt.Errorf("user value field %s has unsupported type %s", field.Name, field.Type)
 	}
 }
 
-// attrInfo returns information on the attribute with the given name.
-func (tinfo *userTypeInfo) attrInfo(name string) attrInfo {
-	info, ok := tinfo.fields[name]
-	if ok {
-		return info
-	}
-	if tinfo.getAttribute == nil {
-		return attrInfo{
-			asString: func(v reflect.Value) string {
-				return ""
-			},
-			asBytes: func(v reflect.Value) []byte {
-				return nil
-			},
-		}
-	}
-	return attrInfo{
-		asString: func(v reflect.Value) string {
-			return tinfo.getAttribute(v, name)
-		},
-		asBytes: func(v reflect.Value) []byte {
-			return []byte(tinfo.getAttribute(v, name))
-		},
-		asSemver: func(v reflect.Value) (*semver.Version, error) {
-			return parseSemver(tinfo.getAttribute(v, name))
-		},
+func isPredefined(attr string) bool {
+	switch attr {
+	case "Identifier", "Email", "Country":
+		return true
 	}
+	return false
 }
 
-func parseSemver(s string) (*semver.Version, error) {
-	if s == "" {
-		return nil, nil
-	}
-	vers, err := semver.Parse(s)
-	if err != nil {
-		return nil, err
-	}
-	return &vers, nil
+func (t *userTypeInfo) getString(v reflect.Value, attr string) (string, bool, error) {
+	info, ok := t.fields[attr]
+	if ok && info.asString != nil {
+		result, converted := info.asString(v)
+		if len(result) == 0 && isPredefined(attr) {
+			return "", false, &userAttrMissingError{attr: attr}
+		}
+		return result, converted, nil
+	} else if t.getAttribute != nil {
+		res := t.getAttribute(v, attr)
+		if res == nil {
+			return "", false, &userAttrMissingError{attr: attr}
+		}
+		switch val := res.(type) {
+		case string:
+			return val, false, nil
+		case []byte:
+			return string(val), false, nil
+		case []string:
+			b, _ := toJson(val)
+			return string(b), false, nil
```

Comment:
String arrays passed to a non-array text comparator should also result in warning 3005.

```suggestion
			return string(b), true, nil
```